### PR TITLE
refactor(dynamite_runtime,nextcloud)!: make DynamiteApiException exte…

### PR DESCRIPTION
**Pull request by Leptopoda**
…nd ClientException.

### CODE REVIEW

File: packages/dynamite/dynamite_runtime/lib/src/dynamite_client.dart [hunk 1]
```diff
@@ -466,7 +442,7 @@ class DynamiteClient {
     if (validStatuses?.contains(response.statusCode) ?? true) {
       return response;
     } else {
-      throw await DynamiteApiException.fromResponse(response);
+      throw DynamiteStatusCodeException(response.statusCode);
```
Comment (provokateurin) [inline packages/dynamite/dynamite_runtime/lib/src/dynamite_client.dart:445]: It's usually very helpful to also see what the body (and headers) are for debugging purposes.

File: packages/nextcloud/lib/src/webdav/client.dart [hunk 2]
```diff
@@ -78,10 +78,8 @@ class WebDavClient {
     final response = await rootClient.httpClient.send(request);
 
     if (response.statusCode >= 300) {
-      throw DynamiteApiException(
+      throw DynamiteStatusCodeException(
         response.statusCode,
-        response.headers,
-        await response.stream.string,
       );
```
Comment (provokateurin) [inline packages/nextcloud/lib/src/webdav/client.dart:83]: Same here

**Review by provokateurin — CHANGES_REQUESTED** [review]

File: packages/dynamite/dynamite_runtime/lib/src/dynamite_client.dart (hunk 1 above)
Comment (Leptopoda) [reply packages/dynamite/dynamite_runtime/lib/src/dynamite_client.dart:445]: But for the body we'd need to await the entire response and there is no reason to have the code wait.

Comment (provokateurin) [reply packages/dynamite/dynamite_runtime/lib/src/dynamite_client.dart:445]: Please add it regardless, otherwise debugging will be a pain.

Comment (Leptopoda) [reply packages/dynamite/dynamite_runtime/lib/src/dynamite_client.dart:445]: Can we make it only do that in debug mode?

Comment (provokateurin) [reply packages/dynamite/dynamite_runtime/lib/src/dynamite_client.dart:445]: Sure

Comment (Leptopoda) [reply packages/dynamite/dynamite_runtime/lib/src/dynamite_client.dart]: How do you like this approach?
I'm not that sure myself.

File: packages/dynamite/dynamite_runtime/lib/src/dynamite_client.dart [hunk 3]
```diff
@@ -466,7 +471,11 @@ class DynamiteClient {
     if (validStatuses?.contains(response.statusCode) ?? true) {
       return response;
     } else {
-      throw await DynamiteApiException.fromResponse(response);
+      if (kDebugMode) {
+        throw await DynamiteStatusCodeException.fromResponse(response);
+      } else {
+        throw DynamiteStatusCodeException(response.statusCode);
+      }
```
Comment (provokateurin) [inline packages/dynamite/dynamite_runtime/lib/src/dynamite_client.dart:478]: Can't you move this logic into the factory? That would make it easier and impossible to make a mistake when only using one of them

Comment (Leptopoda) [reply packages/dynamite/dynamite_runtime/lib/src/dynamite_client.dart:478]: But the method is asynchronous. So the event loop would still be interrupted for no reason.
I'll think about a better solution. Now that this is only used in debug mode we can just block the execution by synchronously waiting.

Comment (provokateurin) [reply packages/dynamite/dynamite_runtime/lib/src/dynamite_client.dart:478]: Ah I see

**Review by provokateurin — APPROVED** [review]

**Pull request merged**